### PR TITLE
[Database] Build Database CommInterface

### LINKED ISSUE #71: [Database] Build and Test Database Production Classes
**What is the use case? Please describe.**
For this sprint task, you'll be building and testing all of the classes in the `edu.oakland.production.database` package. Note: This issue is similar to #65 , but is tracked differently as helper and production classes have different deliverable dates.

**Sprint Task Setup**

To be done by @camrongrant1 , @kylepoterek , or @MathewYaldo before the rest of the team starts working on it:

1. Edit this issue message to add a checkbox for each class, and assign it to the correct team member from #60 .

**Prerequisites**

Please makes sure that your test plan from #60 for the class you are writing is closed and merged to main before starting this, to ensure that all of the documentation you need for this is in the correct spot.

**Steps**

To be done by each team member for each class they are developing a test plan for:

1. Each team member should watch the last ~18 minutes of @andrewdimmer 's [Test Driven Development Crash Course](https://youtu.be/eSc0aY4GhPs)
2. One person should create a branch off of main called `define-Name_Of_Class` (ex. the `Demo` classes and tests get build on a branch called `build-Demo`). Everyone else in the programming pair or group should then checkout that branch.
3. Using the information in the video, use test driven development to build all the required classes, stubs, and tests. Remember: you've already defined what you need to build at https://github.com/andrewdimmer/csi-3370-software-project/tree/main/docs/uml-class-diagrams/database !
4. Commit your changes with the message `[Database] Built Name_Of_Class`
5. Push your changes
6. Create a pull request with the name `[Database] Built Name_Of_Class` and the rest of the information like description, tags, etc. For reference, the settings on the right hand side should match this issue, with the exception that the assignees on the PR are just those who are pushing code.
7. Once your code is complete, add @andrewdimmer as a reviewer.

**Tasks to Complete**

- [x] Allocate Team Members to Classes - @camrongrant1 , @kylepoterek , or @MathewYaldo
- [ ] Build DatabaseCommInterface - @Ian-Klaus 
- [x] Build DatabaseCommManager - @Endritz 
- [x] Build DatabseGisInterface - @robertsonbrinker 
- [x] Build DatabaseGisManager - @Jake-wq 
- [x] Build DatabasePersistentStorage - @MathewYaldo @letthatbe 

**Additional context**
Note: if for some reason you need to change a public interface or shared data type, reach out to @andrewdimmer who can help you manage making that BREAKING CHANGE.

Note: if you run into merge conflicts, let @andrewdimmer know and he can help you resolve them.

### PR DESCRIPTION
**Pull request by MathewYaldo**
Working on #71

### CODE REVIEW

File: .project [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (andrewdimmer) [inline .project]: This is a IDE config file and should be removed from the PR.

File: src/main/java/edu/oakland/production/database/DatabaseCommInterfaceClass.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package edu.oakland.production.database;
+
+import edu.oakland.helper.admin.LocationDataPoint;
+import edu.oakland.helper.admin.TrackData;
+import edu.oakland.production.database.DatabaseCommManager;
+
+public class DatabaseCommInterfaceClass {
```
Comment (andrewdimmer) [inline src/main/java/edu/oakland/production/database/DatabaseCommInterfaceClass.java]: This should be updated to comply with the naming conventions that were laid out in the UML Class Diagrams. You also need to make sure that you implement the `DatabaseCommInterface` Java interface.
```suggestion
public class DatabaseCommInterfaceImplementation implements DatabaseCommInterface {
```

File: src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+import java.time.LocalDateTime;
+import org.junit.jupiter.api.AfterAll;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+@DisplayName("Database CommInterface Test")
+public class DatabaseCommInterfaceTest {
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java]: For sake of standardization to make it easier to read and search the logs, can we change this to `DatabaseCommInterfaceImplementation Unit Tests`? Thank you!
```suggestion
@DisplayName("DatabaseCommInterfaceImplementation Unit Tests")
public class DatabaseCommInterfaceImplementationTests {
```

File: src/test/java/edu/oakland/test/database/DatabaseCommManagerStub.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package edu.oakland.test.database;
+
+import edu.oakland.helper.admin.LocationDataPoint;
+import edu.oakland.helper.admin.TrackData;
+import edu.oakland.production.database.DatabaseCommManager;
+
+public class DatabaseCommManagerStub implements DatabaseCommManager {
+  int rfid = 10;
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommManagerStub.java]: Let's make sure that this is private!
```suggestion
  private int rfid;
```

File: src/test/java/edu/oakland/test/database/DatabaseCommManagerStub.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  LocationDataPoint ldp = null;
+
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommManagerStub.java:17]: Can we add a way to set the RFID number? One way to do it is via constructor (as suggested), or you could also do this via a setter.
```suggestion

  public DatabaseCommManagerStub(int rfid) {
    this.rfid = rfid;
  }

```

File: src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+  @Test
+  @DisplayName("rfid going in is Rfid coming Out")
+  void rfidInIsRfidOut() {
+    int i = 10;
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java]: Let's make this random to increase test robustness:
```suggestion
    int i = (int) (Math.random() * 100);
```

File: src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    DatabaseCommManagerStub dcm = new DatabaseCommManagerStub();
+    assertEquals(i, dcm.passGetRfidRequest());
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java]: For test classes, make sure that you're running the test on the class you want to test. Below is an example of how to use the stub in relation to the class you want to test.
```suggestion
    DatabaseCommManagerStub dcm = new DatabaseCommManagerStub(i);
    DatabaseCommInterface dci = new DatabaseCommInterfaceImpementation(dcm);
    assertEquals(i, dci.receiveGetRfidRequest());
```

File: src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    DatabaseCommManagerStub dcm = new DatabaseCommManagerStub();
+
+    float lat = (float) (Math.random() * 360 - 180);
+    float lng = (float) (Math.random() * 360 - 180);
+    LocationDataPoint ldp = new LocationDataPoint(lat, lng,
+        LocalDateTime.of((int) (Math.random() * 50 + 1970),
+        (int) (Math.random() * 12 + 1),
+        (int) (Math.random() * 28 + 1), (int) (Math.random() * 24),
+        (int) (Math.random() * 60)));
+    ldp = null;
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java]: We shouldn't be setting what we just randomly generated to `null`. `null` is an edge case that should be handled separately, and should not be used as a default test case. Instead, let's pass our randomly generated `LocationDataPoint` to the the `DatabaseCommManagerStub`. You may need to add a setter to the stub to do this.

File: src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    float lat = (float) (Math.random() * 360 - 180);
+    float lng = (float) (Math.random() * 360 - 180);
+    LocationDataPoint ldp = new LocationDataPoint(lat, lng,
+        LocalDateTime.of((int) (Math.random() * 50 + 1970),
+        (int) (Math.random() * 12 + 1),
+        (int) (Math.random() * 28 + 1), (int) (Math.random() * 24),
+        (int) (Math.random() * 60)));
+    ldp = null;    
+    int i = (int) Math.random();    
+    assertEquals(ldp, dcm.passGetLocationDataPointRequest(i));
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java]:
```suggestion
    assertEquals(ldp, dci.receiveGetLocationDataPointRequest(i));
```

File: src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+  }
+  
+  @Test
+  @DisplayName("Offset matches TrackData")
+  void offsetMatchTrackData() {
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java:53]: Can you model this test off of the ones above it?

File: src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+    DatabaseCommManagerStub dcm = new DatabaseCommManagerStub();
+
+    TrackData td = new TrackData(generateRandomLocationDataPointsArray(10));
+    int i = (int) (Math.random() * 7);
+    assertEquals(null, dcm.passGetTrackDataRequest(i));  
+  }
+  
+  @Test
+  @DisplayName("TrackData In is same as TrackData Out")
+  void trackDataInIsTrackDataOut() {
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java:65]: Can you model this test off of the ones above it?

File: src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+    td = null;
+    int i = (int) Math.random();
+    assertEquals(td, dcm.passGetTrackDataRequest(i));
+
+
+  }
+
+  @Test
+  @DisplayName("Mode in is mode out")
+  void modeInIsModeOut() {
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java:79]: Can you model this test off of the ones above it?

**Review by andrewdimmer — CHANGES_REQUESTED** [review]
I've made a number of suggestions to help you get started fixing the testing. If you want to go through and accept my suggestions directly, feel free to. You can also accept as many or as few as you'd like in a batch from https://github.com/andrewdimmer/csi-3370-software-project/pull/138/files, by adding each suggestion you want to add to a batch, then committing the whole batch all at once from the green commit button at the top of the page.

Note that my suggestions are just fixing the lines they are on; you'll likely need to fix additional things like imports or references on other lines to get everything to compile again. They should provide a baseline to model how to do the first two tests though, so you can apply the same techniques to the last two. Let me know if you have any questions, and I'd be happy to provide more info if you need it!

**Review by kylepoterek — APPROVED** [review]

File: src/test/java/edu/oakland/test/database/DatabaseCommManagerStub.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+import edu.oakland.helper.admin.LocationDataPoint;
+import edu.oakland.helper.admin.TrackData;
+import edu.oakland.production.database.DatabaseCommManager;
+
+public class DatabaseCommManagerStub implements DatabaseCommManager {
+  private int rfid;
+  LocationDataPoint ldp = null;
+  int index = 0;
+  TrackData td = null;
+  String mode = "";
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommManagerStub.java]:
```suggestion
  private int rfid;
  private LocationDataPoint ldp = null;
  private int index = 0;
  private TrackData td = null;
  private String mode = "";
```

File: src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+  
+  @Test
+  @DisplayName("TrackData In is same as TrackData Out")
+  void trackDataInIsTrackDataOut() {
+
+    int i = (int) Math.random();
+    DatabaseCommManagerStub dcm = new DatabaseCommManagerStub(i);
+    DatabaseCommInterfaceImplementation dci = new DatabaseCommInterfaceImplementation(dcm);
+    TrackData td = new TrackData(generateRandomLocationDataPointsArray(10));
+    td = null;
```
Comment (andrewdimmer) [inline src/test/java/edu/oakland/test/database/DatabaseCommInterfaceTest.java]: Make sure you're not setting this to null, null is an edge case, not a normal case we want to handle.
```suggestion
```

**Review by andrewdimmer — CHANGES_REQUESTED** [review]
Few tiny changes, but once you take a look at those you are set to merge!

**Review by andrewdimmer — APPROVED** [review]
Just update the branch and you're all set to merge!
